### PR TITLE
ROX-30860: Limit parallel port use by processes 

### PR DESCRIPTION
**Pull request by vikin91**
## Description

Introduced configurable port reuse probability to limit unrealistic endpoint sharing behavior in fake workload generation.

**Key Changes:**
- **Added `OriginatorCache`**: Implements probabilistic endpoint-to-originator caching with configurable reuse probability
- **Added `openPortReuseProbability` field**: Configurable parameter (default 0.05) controlling how often different processes reuse the same IP:port without closing endpoints
- **Updated all workload configurations**: Set realistic 5% port reuse probability across all test scenarios
- **Enhanced network endpoint generation**: Uses consistent originator caching instead of random assignment

**Why**: Previous behavior (100% port reuse in releases ≤4.8) was unrealistic and caused memory pressure in Sensor's enrichment pipeline. Multiple processes listening on the same IP:port without proper endpoint closure created excessive deduplication overhead, as Sensor would retain stale entries indefinitely when no close messages were sent between different originators on the same endpoint.

The new approach simulates realistic container behavior where processes typically bind consistently to endpoints (95% default) while allowing occasional port reuse scenarios (5% for restarts/takeovers).

**Impact**: 
- Reduces memory pressure in test environments while maintaining test coverage for edge cases.
- Long running cluster measurements for 4.9 cannot be directly compared against versions ≤4.8 as the workload profile has changed.

## User-facing documentation

- [x] [CHANGELOG.md](https://github.com/stackrox/stackrox/blob/master/CHANGELOG.md) is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

## Testing and quality

- [x] the change is production ready: the change is [GA](https://github.com/stackrox/stackrox/blob/master/PR_GA.md), or otherwise the functionality is gated by a [feature flag](https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md)
- [x] CI results are [inspected](https://docs.google.com/document/d/1d5ga073jkv4CO1kAJqp8MPGpC6E1bwyrCGZ7S5wKg3w/edit?tab=t.0#heading=h.w4ercgtcg0xp)

### Automated testing

- [x] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

### How I validated my change

- [x] CI and added unit tests
- [x] Manual run on the long-running cluster

The workload runs fine on the long running cluster. 
Sensor is crashing every 12 hours on 8GB of memory (due to OOMKill), which is a known pattern, but that is not caused by this PR, but by the: (1) bug in the memory management that is currently on master, (2) giving Sensor only 8GB of memory for the experiment.

The following chart confirms the rates of objects being generated by fake workflows
Screenshot 2025-09-23 at 10 39 33

This chart shows that the processesListening are being processed and sent to Central
Screenshot 2025-09-23 at 10 46 02

### CODE REVIEW

**Comment by vikin91** [comment]
This change is part of the following stack:

- #16629
    - #16628 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: sensor/kubernetes/fake/flows.go [hunk 1]
```diff
@@ -2,18 +2,92 @@ package fake
...
+// without a 'close' message in between is not a realistic scenario and occurs very rarely in production
+// (practically only when a process deliberately abuses the same IP and Port in parallel to a different process).
+//
+// Setting the portSharingProbability to an unrealistic value (anything higher than 0.05) would cause additional memory-pressure
+// in Sensor enrichment pipeline because the deduping key for processes contains the IP, Port and the originator.
+// Note that if Sensor sees an open endpoint for <container1, 1.1.1.1:80, nginx> and then another open endpoint for
+// <container1, 1.1.1.1:80, apache2>, then Sensor will keep the nginx-entry forever, as there was no 'close' message in between.
+//
+// The probability logic is explicit and configurable for different testing scenarios.
+func (oc *OriginatorCache) GetOrSetOriginator(endpointKey string, containerID string, openPortReuseProbability float32, processPool *ProcessPool) *storage.NetworkProcessUniqueKey {
```
[the remark this answers is not in the document]
Comment (vikin91) [reply sensor/kubernetes/fake/flows.go]: Done [f82abc3](https://github.com/stackrox/stackrox/pull/16628/commits/f82abc31a7f4a9bcddbea40f52d7afaeabf46f60)

File: sensor/kubernetes/fake/flows.go [hunk 2]
```diff
@@ -2,18 +2,99 @@ package fake
...
+// Note that generating multiple 'open' endpoints with the same IP and Port but different originators
+// without a 'close' message in between is not a realistic scenario and occurs very rarely in production
+// (practically only when a process deliberately abuses the same IP and Port in parallel to a different process).
+//
+// Setting the portSharingProbability to an unrealistic value (anything higher than 0.05) would cause additional memory-pressure
+// in Sensor enrichment pipeline because the deduping key for processes contains the IP, Port and the originator.
+// Note that if Sensor sees an open endpoint for <container1, 1.1.1.1:80, nginx> and then another open endpoint for
+// <container1, 1.1.1.1:80, apache2>, then Sensor will keep the nginx-entry forever, as there was no 'close' message in between.
+//
+// The probability logic is explicit and configurable for differ{ent testing scenarios.
```
Comment (JoukoVirtanen) [inline sensor/kubernetes/fake/flows.go]:
```suggestion
// The probability logic is explicit and configurable for different testing scenarios.
```

File: sensor/kubernetes/fake/flows.go [hunk 3]
```diff
@@ -2,18 +2,99 @@ package fake
...
+func (oc *OriginatorCache) GetOrSetOriginator(endpointKey string, containerID string, openPortReuseProbability float64, processPool *ProcessPool) *storage.NetworkProcessUniqueKey {
+	// Ensure that the probability is between 0.0 and 1.0.
+	prob := math.Min(1.0, math.Max(0.0, openPortReuseProbability))
+	if openPortReuseProbability < 0.0 || openPortReuseProbability > 1.0 {
```
Comment (JoukoVirtanen) [inline sensor/kubernetes/fake/flows.go]: Maybe this check can be done just once when the configuration is parsed. Then you can also simplify the line above this one.

File: sensor/kubernetes/fake/workload.go [hunk 4]
```diff
@@ -45,6 +45,10 @@ type NetworkWorkload struct {
 	FlowInterval              time.Duration `yaml:"flowInterval"`
 	BatchSize                 int           `yaml:"batchSize"`
 	GenerateUnclosedEndpoints bool          `yaml:"generateUnclosedEndpoints"`
+	// OpenPortReuseProbability is the probability of reusing an existing open endpoint
+	// by a different process without closing the endpoint.
+	// In releases 4.8 and older, this was not configurable and was always set to1.0.
```
Comment (JoukoVirtanen) [inline sensor/kubernetes/fake/workload.go]:
```suggestion
	// In releases 4.8 and older, this was not configurable and was always set to 1.0.
```

File: sensor/kubernetes/fake/flows.go (hunk 3 above)
Comment (vikin91) [reply sensor/kubernetes/fake/flows.go]: Done in [e7cdf92](https://github.com/stackrox/stackrox/pull/16628/commits/e7cdf921710ea0a1d451c018afd828143a378225)

File: sensor/kubernetes/fake/flows.go (hunk 2 above)
Comment (vikin91) [reply sensor/kubernetes/fake/flows.go]: Done [e7cdf92](https://github.com/stackrox/stackrox/pull/16628/commits/e7cdf921710ea0a1d451c018afd828143a378225)

File: sensor/kubernetes/fake/workload.go (hunk 4 above)
Comment (vikin91) [reply sensor/kubernetes/fake/workload.go]: Done [e7cdf92](https://github.com/stackrox/stackrox/pull/16628/commits/e7cdf921710ea0a1d451c018afd828143a378225)

File: sensor/kubernetes/fake/flows.go (hunk 3 above)
Comment (JoukoVirtanen) [reply sensor/kubernetes/fake/flows.go]: [nit] The function name `validateWorkload` implies that there are many checks in it. I don't think that has to be done now. Maybe a comment that more checks should be added to it in the future is good enough for now.

Comment (JoukoVirtanen) [reply sensor/kubernetes/fake/flows.go]: [nit] I was thinking that `openPortReuseProbability` could also be set to 0 or 1.0 when the validation is done, and then instead of using `prob` you could just use `openPortReuseProbability`.

Comment (vikin91) [reply sensor/kubernetes/fake/flows.go]: Oh, so like change it directly in the `workload` object? Yeah, I can do that

Comment (vikin91) [reply sensor/kubernetes/fake/flows.go]: Here it is: [ddb5b45](https://github.com/stackrox/stackrox/pull/16628/commits/ddb5b450d06b9d116b0f3e0d2ebdc503abb45323)

**Comment by vikin91** [comment]
I still haven't done the long run on a cluster, but I will trigger it tomorrow.
The expectation is that the load on Sensor and Central will be lower.

Edit: running on `pr-09-19-long`

**Comment by vikin91** [comment]
@JoukoVirtanen PTAL again, I have observations from the long-running cluster that workload generation works fine.

**Review by JoukoVirtanen — APPROVED** [review]
LGTM!
